File: phases/04-algorithms-and-complexity/03-sorting-ii-merge-quick-and-quickselect/code/main.py
"""
Sorting II — Merge, Quick (and Quickselect)
Phase 04 — Algorithms & Complexity Analysis

From-scratch implementations with step counting and pivot strategy comparison.
"""

import random
import time
from typing import Callable


# ---------------------------------------------------------------------------
# Merge Sort
# ---------------------------------------------------------------------------

class MergeSortCounter:
    """Merge sort with comparison and move counters."""

    def __init__(self) -> None:
        self.comparisons = 0
        self.moves = 0

    def sort(self, arr: list[int]) -> list[int]:
        self.comparisons = 0
        self.moves = 0
        return self._merge_sort(arr)

    def _merge_sort(self, arr: list[int]) -> list[int]:
        if len(arr) <= 1:
            return arr
        mid = len(arr) // 2
        left = self._merge_sort(arr[:mid])
        right = self._merge_sort(arr[mid:])
        return self._merge(left, right)

    def _merge(self, left: list[int], right: list[int]) -> list[int]:
        result = []
        i = j = 0
        while i < len(left) and j < len(right):
            self.comparisons += 1
            if left[i] <= right[j]:
                result.append(left[i])
                self.moves += 1
                i += 1
            else:
                result.append(right[j])
                self.moves += 1
                j += 1
        while i < len(left):
            result.append(left[i])
            self.moves += 1
            i += 1
        while j < len(right):
            result.append(right[j])
            self.moves += 1
            j += 1
        return result


# ---------------------------------------------------------------------------
# Quicksort with pivot strategies
# ---------------------------------------------------------------------------

class QuickSortCounter:
    """Quicksort with comparison and swap counters, pluggable pivot strategy."""

    def __init__(self, pivot_strategy: str = "median3") -> None:
        self.strategy = pivot_strategy
        self.comparisons = 0
        self.swaps = 0
        self._rng = random.Random(42)  # deterministic for benchmarking

    def sort(self, arr: list[int]) -> None:
        self.comparisons = 0
        self.swaps = 0
        self._qs(arr, 0, len(arr) - 1)

    def _qs(self, arr: list[int], lo: int, hi: int) -> None:
        if lo >= hi:
            return
        pivot_idx = self._choose_pivot(arr, lo, hi)
        arr[lo], arr[pivot_idx] = arr[pivot_idx], arr[lo]
        self.swaps += 1
        p = self._partition(arr, lo, hi)
        self._qs(arr, lo, p - 1)
        self._qs(arr, p + 1, hi)

    def _choose_pivot(self, arr: list[int], lo: int, hi: int) -> int:
        if self.strategy == "first":
            return lo
        if self.strategy == "random":
            return self._rng.randint(lo, hi)
        if self.strategy == "median3":
            mid = (lo + hi) // 2
            a, b, c = arr[lo], arr[mid], arr[hi]
            if a <= b <= c or c <= b <= a:
                return mid
            if b <= a <= c or c <= a <= b:
                return lo
            return hi
        raise ValueError(f"Unknown strategy: {self.strategy}")

    def _partition(self, arr: list[int], lo: int, hi: int) -> int:
        """Lomuto partition."""
        pivot = arr[lo]
        i = lo + 1
        for j in range(lo + 1, hi + 1):
            self.comparisons += 1
            if arr[j] < pivot:
                arr[i], arr[j] = arr[j], arr[i]
                self.swaps += 1
                i += 1
        arr[lo], arr[i - 1] = arr[i - 1], arr[lo]
        self.swaps += 1
        return i - 1


# ---------------------------------------------------------------------------
# Quicksort with 3-way partition (Dutch National Flag)
# ---------------------------------------------------------------------------

class QuickSort3Way:
    """Quicksort with 3-way partition — efficient when many duplicates exist."""

    def __init__(self) -> None:
        self.comparisons = 0
        self.swaps = 0

    def sort(self, arr: list[int]) -> None:
        self.comparisons = 0
        self.swaps = 0
        self._qs3(arr, 0, len(arr) - 1)

    def _qs3(self, arr: list[int], lo: int, hi: int) -> None:
        if lo >= hi:
            return
        lt, gt = self._partition3(arr, lo, hi)
        self._qs3(arr, lo, lt - 1)
        self._qs3(arr, gt + 1, hi)

    def _partition3(self, arr: list[int], lo: int, hi: int) -> tuple[int, int]:
        pivot = arr[lo]
        lt = lo       # arr[lo..lt-1] < pivot
        gt = hi       # arr[gt+1..hi] > pivot
        i = lo        # arr[lt..i-1] == pivot
        while i <= gt:
            self.comparisons += 1
            if arr[i] < pivot:
                arr[lt], arr[i] = arr[i], arr[lt]
                self.swaps += 1
                lt += 1
                i += 1
            elif arr[i] > pivot:
                arr[i], arr[gt] = arr[gt], arr[i]
                self.swaps += 1
                gt -= 1
            else:
                i += 1
        return lt, gt


# ---------------------------------------------------------------------------
# Quickselect
# ---------------------------------------------------------------------------

def quickselect(arr: list[int], k: int) -> int:
    """Find the k-th smallest element (0-indexed). Average O(n)."""
    if not 0 <= k < len(arr):
        raise IndexError(f"k={k} out of range for length {len(arr)}")
    lo, hi = 0, len(arr) - 1
    rng = random.Random(42)
    while lo < hi:
        pivot_idx = rng.randint(lo, hi)
        arr[lo], arr[pivot_idx] = arr[pivot_idx], arr[lo]
        pivot = arr[lo]
        i = lo + 1
        for j in range(lo + 1, hi + 1):
            if arr[j] < pivot:
                arr[i], arr[j] = arr[j], arr[i]
                i += 1
        arr[lo], arr[i - 1] = arr[i - 1], arr[lo]
        p = i - 1
        if p == k:
            return arr[p]
        elif p < k:
            lo = p + 1
        else:
            hi = p - 1
    return arr[lo]


# ---------------------------------------------------------------------------
# Benchmarking harness
# ---------------------------------------------------------------------------

def benchmark(sort_fn: Callable[[list[int]], None], data: list[int], label: str) -> dict:
    """Time a sort function and return stats."""
    arr = data.copy()
    start = time.perf_counter()
    sort_fn(arr)
    elapsed = time.perf_counter() - start
    return {"label": label, "time_ms": elapsed * 1000, "sorted": arr}


def main() -> None:
    print("=== Sorting II — Merge, Quick, Quickselect ===\n")

    # --- Demo: merge sort ---
    ms = MergeSortCounter()
    data = [38, 27, 43, 3, 9, 82, 10]
    result = ms.sort(data)
    print(f"Merge sort:  {data} -> {result}")
    print(f"  comparisons={ms.comparisons}, moves={ms.moves}\n")

    # --- Demo: quicksort strategies ---
    strategies = ["first", "random", "median3"]
    for s in strategies:
        qs = QuickSortCounter(s)
        arr = data.copy()
        qs.sort(arr)
        print(f"Quicksort ({s:>7}): {arr}  "
              f"comparisons={qs.comparisons}, swaps={qs.swaps}")
    print()

    # --- Demo: 3-way quicksort ---
    qs3 = QuickSort3Way()
    dup_data = [5, 3, 5, 1, 5, 3, 5, 1, 5, 3]
    qs3.sort(dup_data)
    print(f"3-way QS (many dups): {dup_data}  "
          f"comparisons={qs3.comparisons}, swaps={qs3.swaps}\n")

    # --- Demo: quickselect ---
    arr = [38, 27, 43, 3, 9, 82, 10]
    for k in [0, 2, 6]:
        val = quickselect(arr.copy(), k)
        print(f"quickselect(arr, k={k}) = {val}")
    print()

    # --- Benchmark table ---
    sizes = [1000, 5000, 10000]
    results: dict[str, list[float]] = {}
    print(f"{'Algorithm':<25} {'n=1000':>10} {'n=5000':>10} {'n=10000':>10}")
    print("-" * 58)

    for size in sizes:
        random.seed(42)
        data = [random.randint(0, size) for _ in range(size)]

        # Merge sort
        ms = MergeSortCounter()
        start = time.perf_counter()
        ms.sort(data)
        t = (time.perf_counter() - start) * 1000
        results.setdefault("merge_sort", []).append(t)

        for s in strategies:
            qs = QuickSortCounter(s)
            arr = data.copy()
            start = time.perf_counter()
            qs.sort(arr)
            t = (time.perf_counter() - start) * 1000
            key = f"quick_{s}"
            results.setdefault(key, []).append(t)

        # 3-way quicksort
        qs3 = QuickSort3Way()
        arr = data.copy()
        start = time.perf_counter()
        qs3.sort(arr)
        t = (time.perf_counter() - start) * 1000
        results.setdefault("quick_3way", []).append(t)

    for algo, times in results.items():
        print(f"{algo:<25} {times[0]:>9.2f}ms {times[1]:>9.2f}ms {times[2]:>9.2f}ms")
    print()

    # --- Stability demo ---
    print("Stability demo:")
    pairs = [(3, 'a'), (1, 'b'), (3, 'c'), (1, 'd')]
    # Merge sort — stable
    ms2 = MergeSortCounter()
    sorted_pairs = ms2.sort([p[0] for p in pairs])
    # Reconstruct by original order for equal keys
    print(f"  Input:   {pairs}")
    print(f"  Merge sort preserves order of equal keys: stable ✓")
    print(f"  Quicksort does NOT guarantee this:        stable ?\n")


if __name__ == "__main__":
    main()

File: phases/04-algorithms-and-complexity/03-sorting-ii-merge-quick-and-quickselect/code/main.rs
//! Sorting II — Merge, Quick (and Quickselect)
//! Phase 04 — Algorithms & Complexity Analysis
//!
//! Generics-based merge sort and quicksort implementations with step counting.

use std::fmt::Debug;

// ---------------------------------------------------------------------------
// Merge Sort
// ---------------------------------------------------------------------------

/// Sort a slice in place using merge sort. Returns comparison and move counts.
pub fn merge_sort<T: Ord + Copy>(arr: &mut [T]) -> (usize, usize) {
    let mut comparisons = 0usize;
    let mut moves = 0usize;
    let len = arr.len();
    if len <= 1 {
        return (0, 0);
    }
    let mut buf = vec![arr[0]; len]; // reuse buffer
    merge_sort_inner(arr, &mut buf, 0, len, &mut comparisons, &mut moves);
    (comparisons, moves)
}

fn merge_sort_inner<T: Ord + Copy>(
    arr: &mut [T],
    buf: &mut [T],
    lo: usize,
    hi: usize,
    comparisons: &mut usize,
    moves: &mut usize,
) {
    if hi - lo <= 1 {
        return;
    }
    let mid = lo + (hi - lo) / 2;
    merge_sort_inner(arr, buf, lo, mid, comparisons, moves);
    merge_sort_inner(arr, buf, mid, hi, comparisons, moves);
    // Merge arr[lo..mid] and arr[mid..hi] into buf[lo..hi], then copy back.
    let mut i = lo;
    let mut j = mid;
    let mut k = lo;
    while i < mid && j < hi {
        *comparisons += 1;
        if arr[i] <= arr[j] {
            buf[k] = arr[i];
            *moves += 1;
            i += 1;
        } else {
            buf[k] = arr[j];
            *moves += 1;
            j += 1;
        }
        k += 1;
    }
    while i < mid {
        buf[k] = arr[i];
        *moves += 1;
        i += 1;
        k += 1;
    }
    while j < hi {
        buf[k] = arr[j];
        *moves += 1;
        j += 1;
        k += 1;
    }
    arr[lo..hi].copy_from_slice(&buf[lo..hi]);
}

// ---------------------------------------------------------------------------
// Quicksort
// ---------------------------------------------------------------------------

pub enum PivotStrategy {
    First,
    Random,
    Median3,
}

pub struct QuickSortStats {
    pub comparisons: usize,
    pub swaps: usize,
}

pub fn quick_sort<T: Ord + Copy + Debug>(
    arr: &mut [T],
    strategy: &PivotStrategy,
) -> QuickSortStats {
    let mut stats = QuickSortStats {
        comparisons: 0,
        swaps: 0,
    };
    quick_sort_inner(arr, strategy, &mut stats);
    stats
}

fn quick_sort_inner<T: Ord + Copy>(
    arr: &mut [T],
    strategy: &PivotStrategy,
    stats: &mut QuickSortStats,
) {
    let len = arr.len();
    if len <= 1 {
        return;
    }
    let pivot_idx = choose_pivot(arr, strategy);
    arr.swap(0, pivot_idx);
    stats.swaps += 1;
    let p = partition_lomuto(arr, stats);
    quick_sort_inner(&mut arr[..p], strategy, stats);
    quick_sort_inner(&mut arr[p + 1..], strategy, stats);
}

fn choose_pivot<T: Ord + Copy>(arr: &[T], strategy: &PivotStrategy) -> usize {
    let hi = arr.len() - 1;
    match strategy {
        PivotStrategy::First => 0,
        PivotStrategy::Random => {
            // Simple deterministic-ish for demo; use rand crate in production.
            let mid = hi / 2;
            (arr[0] as usize ^ arr[mid] as usize ^ arr[hi] as usize) % arr.len()
        }
        PivotStrategy::Median3 => {
            let lo = 0;
            let mid = hi / 2;
            let a = arr[lo];
            let b = arr[mid];
            let c = arr[hi];
            if (a <= b && b <= c) || (c <= b && b <= a) {
                mid
            } else if (b <= a && a <= c) || (c <= a && a <= b) {
                lo
            } else {
                hi
            }
        }
    }
}

fn partition_lomuto<T: Ord + Copy>(arr: &mut [T], stats: &mut QuickSortStats) -> usize {
    let pivot = arr[0];
    let mut i = 1;
    for j in 1..arr.len() {
        stats.comparisons += 1;
        if arr[j] < pivot {
            arr.swap(i, j);
            stats.swaps += 1;
            i += 1;
        }
    }
    arr.swap(0, i - 1);
    stats.swaps += 1;
    i - 1
}

// ---------------------------------------------------------------------------
// Quickselect
// ---------------------------------------------------------------------------

pub fn quickselect<T: Ord + Copy>(arr: &mut [T], k: usize) -> T {
    assert!(k < arr.len(), "k out of bounds");
    let mut lo = 0usize;
    let mut hi = arr.len() - 1;
    while lo < hi {
        let mid = lo + (hi - lo) / 2;
        // Median-of-3 pivot selection
        let pivot_idx = {
            let (a, b, c) = (arr[lo], arr[mid], arr[hi]);
            if (a <= b && b <= c) || (c <= b && b <= a) {
                mid
            } else if (b <= a && a <= c) || (c <= a && a <= b) {
                lo
            } else {
                hi
            }
        };
        arr.swap(lo, pivot_idx);
        let pivot = arr[lo];
        let mut i = lo + 1;
        for j in lo + 1..=hi {
            if arr[j] < pivot {
                arr.swap(i, j);
                i += 1;
            }
        }
        arr.swap(lo, i - 1);
        let p = i - 1;
        if p == k {
            return arr[p];
        } else if p < k {
            lo = p + 1;
        } else {
            hi = p - 1;
        }
    }
    arr[lo]
}

// ---------------------------------------------------------------------------
// Main — demos and benchmarks
// ---------------------------------------------------------------------------

fn main() {
    println!("=== Sorting II — Merge, Quick, Quickselect ===\n");

    // --- Merge sort demo ---
    let mut data = vec![38, 27, 43, 3, 9, 82, 10];
    let (cmp, mov) = merge_sort(&mut data);
    println!("Merge sort:  {:?}", data);
    println!("  comparisons={cmp}, moves={mov}\n");

    // --- Quicksort strategies demo ---
    for (name, strat) in [
        ("first", PivotStrategy::First),
        ("median3", PivotStrategy::Median3),
    ] {
        let mut arr = vec![38, 27, 43, 3, 9, 82, 10];
        let stats = quick_sort(&mut arr, &strat);
        println!(
            "Quicksort ({:>7}): {:?}  comparisons={}, swaps={}",
            name, arr, stats.comparisons, stats.swaps
        );
    }

    // --- Quickselect demo ---
    let mut arr = vec![38, 27, 43, 3, 9, 82, 10];
    for k in [0, 2, 6] {
        let val = quickselect(&mut arr.clone(), k);
        println!("quickselect(arr, k={k}) = {val}");
    }
    println!();

    // --- Benchmark ---
    println!(
        "{:<25} {:>10} {:>10} {:>10}",
        "Algorithm", "n=1000", "n=5000", "n=10000"
    );
    println!("{}", "-".repeat(58));

    let sizes = [1000, 5000, 10000];
    let mut merge_times = Vec::new();
    let mut qs_times = Vec::new();

    for &size in &sizes {
        let mut data: Vec<i32> = (0..size).map(|i| (i * 7 + 13) % size as i32).collect();

        // Merge sort
        let mut d1 = data.clone();
        let start = std::time::Instant::now();
        merge_sort(&mut d1);
        merge_times.push(start.elapsed().as_secs_f64() * 1000.0);

        // Quicksort (median3)
        let mut d2 = data.clone();
        let start = std::time::Instant::now();
        quick_sort(&mut d2, &PivotStrategy::Median3);
        qs_times.push(start.elapsed().as_secs_f64() * 1000.0);
    }

    println!(
        "{:<25} {:>9.2}ms {:>9.2}ms {:>9.2}ms", "merge_sort",
        merge_times[0], merge_times[1], merge_times[2]
    );
    println!(
        "{:<25} {:>9.2}ms {:>9.2}ms {:>9.2}ms", "quick_median3",
        qs_times[0], qs_times[1], qs_times[2]
    );
}
